ports/esp32/machine_i2c.c

```diff
@@ -49,92 +49,159 @@
     };
-    i2c_param_config(self->port, &conf);
-    int timeout = I2C_SCLK_FREQ / 1000000 * timeout_us;
-    i2c_set_timeout(self->port, (timeout > I2C_LL_MAX_TIMEOUT) ? I2C_LL_MAX_TIMEOUT : timeout);
-    i2c_driver_install(self->port, I2C_MODE_MASTER, 0, 0, 0);
+    ESP_ERROR_CHECK(i2c_new_master_bus(&bus_cfg, &self->bus_handle));
+
+    // 3. Add a device (placeholder address; will be changed dynamically later)
+    i2c_device_config_t dev_cfg = {
+        .dev_addr_length = I2C_ADDR_BIT_LEN_7,
+        .device_address = 0x00,               // Placeholder
+        .scl_speed_hz = freq,
+    };
+    ESP_ERROR_CHECK(i2c_master_bus_add_device(self->bus_handle, &dev_cfg, &self->dev_handle));
 }
 
 int machine_hw_i2c_transfer(mp_obj_base_t *self_in, uint16_t addr, size_t n, mp_machine_i2c_buf_t *bufs, unsigned int flags) {
     machine_hw_i2c_obj_t *self = MP_OBJ_TO_PTR(self_in);
 
-    i2c_cmd_handle_t cmd = i2c_cmd_link_create();
+    /* 0. Probe the address to see if any device responds */
+    esp_err_t err = i2c_master_probe(self->bus_handle, addr, 1000);
+    if (err != ESP_OK) {
+        return -MP_ENODEV;          /* No device at address, return immediately */
+    }
+    /* 1. Create a temporary device handle for this transaction */
+    i2c_device_config_t dev_cfg = {
+        .dev_addr_length = I2C_ADDR_BIT_LEN_7,
+        .device_address = addr,
+        .scl_speed_hz = 100000,   /* Use bus frequency */
+    };
+    i2c_master_dev_handle_t dev_handle;
+    err = i2c_master_bus_add_device(self->bus_handle, &dev_cfg, &dev_handle);
+    if (err != ESP_OK) {
+        return -MP_ENODEV;
+    }
+
     int data_len = 0;
 
+    /* 2. If WRITE1 segment exists, perform the write first */
     if (flags & MP_MACHINE_I2C_FLAG_WRITE1) {
-        i2c_master_start(cmd);
-        i2c_master_write_byte(cmd, addr << 1, true);
-        i2c_master_write(cmd, bufs->buf, bufs->len, true);
+        if (bufs->len) {
+            err = i2c_master_transmit(dev_handle, bufs->buf, bufs->len, 1000);                 /* Block with 1 s timeout */
+            if (err != ESP_OK) {
+                goto cleanup;
+            }
+        }
         data_len += bufs->len;
         --n;
         ++bufs;
     }
+    if (flags & MP_MACHINE_I2C_FLAG_READ) {
+        /* 3. Main loop: remaining segments */
+        for (; n--; ++bufs) {
+            if (bufs->len == 0) {
+                continue;
+            }
+            err = i2c_master_receive(dev_handle, bufs->buf, bufs->len, 1000);
+            if (err != ESP_OK) {
+                break;
+            }
+
+            data_len += bufs->len;
+        }
+    } else {
+        // Write operation logic
+        size_t total_len = 0;
+        mp_machine_i2c_buf_t *original_bufs = bufs;  // Save original pointer
+        size_t yuann = n;
 
-    i2c_master_start(cmd);
-    i2c_master_write_byte(cmd, addr << 1 | (flags & MP_MACHINE_I2C_FLAG_READ), true);
+        // Calculate total length
```

Comment:
Is it possible to use i2c_master_multi_buffer_transmit() instead of copying the segments into a single buffer?